content/recipes/swagger.md

```diff
@@ -77,75 +73,179 @@ async create(@Body() createCatDto: CreateCatDto) {
 
 ```typescript
-export const ApiModelProperty: (metadata?: {
-  description?: string;
-  required?: boolean;
-  type?: any;
-  isArray?: boolean;
-  collectionFormat?: string;
-  default?: any;
-  enum?: SwaggerEnumType;
-  format?: string;
-  multipleOf?: number;
-  maximum?: number;
-  exclusiveMaximum?: number;
-  minimum?: number;
-  exclusiveMinimum?: number;
-  maxLength?: number;
-  minLength?: number;
-  pattern?: string;
-  maxItems?: number;
-  minItems?: number;
-  uniqueItems?: boolean;
-  maxProperties?: number;
-  minProperties?: number;
-  readOnly?: boolean;
-  xml?: any;
-  example?: any;
-}) => PropertyDecorator;
-```
-
-> warning **Hint** There's an `@ApiModelPropertyOptional()` shortcut decorator which helps to avoid continuous typing `{{"@ApiModelProperty({ required: false })"}}`.
-
-Thanks to that we can simply set the **default** value, determine whether the property is required or explicitly set the type.
+@ApiProperty({
+  description: 'The age of a cat',
+  min: 1,
+  default: 1,
+})
+age: number;
+```
+
+> info **Hint** Instead of explicitly typing `{{"@ApiProperty({ required: false })"}}` you can use `@ApiPropertyOptional()` short-hand decorator.
+
+In order to explicitly set the type of the property, use `type` key:
+
+```typescript
+@ApiProperty({
+  type: Number,
+})
+age: number;
+```
+
+#### Enums
+
+To be able for `SwaggerModule` to identify an `enum`, we have to manually set the `enum` property on `@ApiProperty` with an array of values.
+
+```typescript
+@ApiProperty({ enum: ['Admin', 'Moderator', 'User']})
+role: UserRole;
+```
+
+You can also define a real `UserRole` enum instead as following:
+
+```typescript
+export enum UserRole {
+  Admin = 'Admin',
+  Moderator = 'Moderator',
+  User = 'User',
+}
+```
+
+Enums can be used by itself with the `@Query()` parameter decorator in combination with the `@ApiQuery()` decorator.
+
+```typescript
+@ApiQuery({ name: 'role', enum: UserRole })
+async filterByRole(@Query('role') role: UserRole = UserRole.User) {}
+```
+
+<figure><img src="/assets/enum_query.gif" /></figure>
+
+With `isArray` set to **true**, the `enum` can be selected as a **multi-select**:
+
+<figure><img src="/assets/enum_query_array.gif" /></figure>
+
+#### Arrays
+
+We have to manually indicate a type when the property is an array:
+
+```typescript
+@ApiProperty({ type: [String] })
+names: string[];
+```
+
+Simply put your type as the first element of an array (as shown above) or set an `isArray` property to `true`.
+
+#### Circular dependencies
+
+When you have cross-references between your classes, ensure to use lazy functions in order to provide a module with a type information lazily.
+
+```typescript
+@ApiProperty({ type: () => Node })
+node: Node;
+```
+
+#### Generics and interfaces
+
+TypeScript does not store any metadata about generics and interfaces. Hence, if you use them to define your DTOs, we may not be able to properly generate model definitons at the runtime.
```

Comment:
TypeScript does not store any metadata about generics and interfaces. Hence, if you use them to define your DTOs, we may not be able to properly generate model definitons at the runtime.

->

Since TypeScript does not store metadata about generics or interfaces, when you use them in your DTOs, `SwaggerModule` may not be able to properly generate model definitions at runtime.